orebfuscator-plugin/src/main/java/net/imprex/orebfuscator/injector/OrebfuscatorInjector.java

```diff
@@ -0,0 +1,168 @@
+package net.imprex.orebfuscator.injector;
+
+import java.lang.reflect.Modifier;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.bukkit.entity.Player;
+
+import com.comphenix.protocol.PacketType;
+import com.comphenix.protocol.events.PacketContainer;
+import com.comphenix.protocol.reflect.FuzzyReflection;
+import com.comphenix.protocol.reflect.accessors.Accessors;
+import com.comphenix.protocol.reflect.fuzzy.FuzzyFieldContract;
+import com.comphenix.protocol.utility.MinecraftFields;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelPipeline;
+import io.netty.util.concurrent.Promise;
+import net.imprex.orebfuscator.Orebfuscator;
+import net.imprex.orebfuscator.chunk.ChunkStruct;
+import net.imprex.orebfuscator.config.AdvancedConfig;
+import net.imprex.orebfuscator.config.OrebfuscatorConfig;
+import net.imprex.orebfuscator.obfuscation.ObfuscationResult;
+import net.imprex.orebfuscator.obfuscation.ObfuscationSystem;
+import net.imprex.orebfuscator.player.OrebfuscatorPlayer;
+import net.imprex.orebfuscator.player.OrebfuscatorPlayerMap;
+import net.imprex.orebfuscator.util.BlockPos;
+import net.imprex.orebfuscator.util.OFCLogger;
+import net.imprex.orebfuscator.util.PermissionUtil;
+
+public class OrebfuscatorInjector {
+
+	private static final String OUTBOUND_HANDLER_NAME = "orebfuscator_outbound";
+
+	private final OrebfuscatorConfig config;
+	private final OrebfuscatorPlayerMap playerMap;
+	private final ObfuscationSystem obfuscationSystem;
+
+	private final Player player;
+	private final Channel channel;
+
+	private final Queue<Long> delayPackets = new LinkedList<Long>();
+
+	public OrebfuscatorInjector(Orebfuscator orebfuscator, Player player) {
+		this.config = orebfuscator.getOrebfuscatorConfig();
+		this.playerMap = orebfuscator.getPlayerMap();
+		this.obfuscationSystem = orebfuscator.getObfuscationSystem();
+
+		Object networkManager = MinecraftFields.getNetworkManager(player);
+		var channelAccessor = Accessors.getFieldAccessor(FuzzyReflection.fromObject(networkManager, true)
+				.getField(FuzzyFieldContract.newBuilder()
+				.typeExact(Channel.class)
+				.banModifier(Modifier.STATIC)
+				.build()));
+
+		this.player = player;
+		this.channel = (Channel) channelAccessor.get(networkManager);
+
+		ChannelPipeline pipeline = this.channel.pipeline();
+
+		String encoderName = pipeline.get("outbound_config") != null
+				? "outbound_config"
+				: "encoder";
+
+		if (pipeline.context(OUTBOUND_HANDLER_NAME) == null) {
+			pipeline.addAfter(encoderName, OUTBOUND_HANDLER_NAME, new AsyncOutboundPacketHandler(this));
+		}
+	}
+
+	public void uninject() {
+		if (!this.channel.eventLoop().inEventLoop()) {
```

Comment:
Is uninject really should be done in event loop always? On plugin disable i think it should not switch contexts. 
I think nobody is reloading orebfuscator via PlugMan, but for developing purposes it can be done and there is a really low chance that on plugin reloading it may fail to inject because old handler is still present on pipeline.

Reply:
I don't think it would hurt to remove the event-loop context switch so I will do that.